cpp/ql/lib/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -3416,7 +3416,7 @@ private predicate pathSuccPlus(PathNode n1, PathNode n2) = fastTC(pathSucc/2)(n1
  */
 module PathGraph {
   /** Holds if `(a,b)` is an edge in the graph of data flow path explanations. */
-  query predicate edges(PathNode a, PathNode b) { a.getASuccessor() = b and reach(b) }
+  query predicate edges(PathNode a, PathNode b) { a.getASuccessor() = b and reach(a) and reach(b) }
```

Comment:
I wonder if we shouldn't restrict sub path nodes to those that can reach a sink as well. I.e., redefine `retReach` as
```ql
  predicate retReach(PathNode n) {
    exists(PathNodeMid out |
      subpaths(_, _, n, out) and
      reach(out)
    )
    or
    exists(PathNode mid |
      retReach(mid) and
      n.getASuccessor() = mid and
      not subpaths(_, mid, _, _)
    )
  }
```

@aschackmull : What do you think?

Reply:
That makes sense, although I think I'd prefer
```ql
  predicate retReach(PathNode n) {
    exists(PathNodeMid out |
      subpaths(_, _, n, out)
    |
      directReach(out) or retReach(out)
    )
    or
    exists(PathNode mid |
      retReach(mid) and
      n.getASuccessor() = mid and
      not subpaths(_, mid, _, _)
    )
  }
```
to avoid making it mutually recursive with `reach`.

Reply:
We might also want to restrict the final `subpaths` relation if we want to get rid of the `Skipping subpaths tuple with missing node` warnings.

Reply:
Yes, we should certainly also restrict `subpaths`.